var/spack/repos/builtin/packages/silo/package.py

```diff
@@ -27,15 +27,25 @@ class Silo(AutotoolsPackage):
             description='Produce position-independent code (for shared libs)')
     variant('mpi', default=True,
             description='Compile with MPI Compatibility')
+    variant('szip', default=False,
+            description='Enable szip support')
 
-    depends_on('hdf5@:1.10.999', when='@:4.10.2')
-    depends_on('hdf5~mpi', when='~mpi')
     depends_on('mpi', when='+mpi')
-    depends_on('hdf5+mpi', when='+mpi')
+    depends_on('hdf5@:1.10.999+szip', when='@:4.10.2+szip')
+    depends_on('hdf5@:1.10.999~szip', when='@:4.10.2~szip')
+    depends_on('hdf5~mpi+szip', when='~mpi+szip')
+    depends_on('hdf5+mpi+szip', when='+mpi+szip')
+    depends_on('hdf5~mpi~szip', when='~mpi~szip')
+    depends_on('hdf5+mpi~szip', when='+mpi~szip')
     depends_on('qt~framework@4.8:4.9', when='+silex')
     depends_on('libx11', when='+silex')
+    # Xmu dependency is required on Ubuntu 18-20
+    depends_on('libxmu', when='+silex')
+    # Silo configure assumes readline by default, so better add it here (for browser)
```

Comment:
What's the error that you see when `libxmu` isn't explicitly linked in? Are you sure it's an issue with silex and not x11 or qt4?

Also, delete the readline comment.

Reply:
Doing step ax_err_4 from configure manually:
$CXX -o ax_qt_main ax_qt_main.o moc_ax_qt_test.o $QT_LIBS $LIBS
/usr/bin/ld: cannot find -lXmu
collect2: error: ld returned 1 exit status
echo "$CXX -o ax_qt_main ax_qt_main.o moc_ax_qt_test.o $QT_LIBS $LIBS"
g++ -o ax_qt_main ax_qt_main.o moc_ax_qt_test.o -L/home/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/lib -lQtGui -lQtCore  -lSM -lICE  -L/home/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/libx11-1.7.0-wdieafv2ms7kxfkbi43ky6xhspdzg4y4/lib -lX11 -lXext -lXmu -lXt -lXi

Reply:
That checks out -- looks like Silo is explicitly asking for those.  Could you double-check that you're linking against `qt+gui`, and update the `qt` dependency in silo to match? Thanks!

Reply:
I'm not sure to understand what you ask ; silo depends on qt which by default uses the +gui variant.
Would the spec answer your question ?
$ spack spec silo+silex~mpi ^hdf5~mpi+szip
Input spec
--------------------------------
silo~mpi+silex
    ^hdf5~mpi+szip

Concretized
--------------------------------
==> Warning: gcc@7.5.0 cannot build optimized binaries for "zen2". Using best target possible: "zen"
silo@4.10.2%gcc@7.5.0+fortran~mpi+pic+shared+silex patches=7b5a1dc2a0e358e667088d77e7caa780967fa8ea60be89c44986605df9990abe,952d3c95fbeaf32355c1c24d831b6c315c5cf29023874127b0f461855046750c arch=linux-ubuntu18.04-zen
    ^hdf5@1.10.7%gcc@7.5.0~cxx~debug~fortran~hl~java~mpi+pic+shared+szip~threadsafe api=none arch=linux-ubuntu18.04-zen
        ^libszip@2.1.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^zlib@1.2.11%gcc@7.5.0+optimize+pic+shared arch=linux-ubuntu18.04-zen
    ^libx11@1.7.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^inputproto@2.3.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^pkgconf@1.7.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^util-macros@1.19.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^kbproto@1.0.7%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libxcb@1.14%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libpthread-stubs@0.4%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libxau@1.0.8%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^xproto@7.0.31%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libxdmcp@1.1.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^libbsd@0.10.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^xcb-proto@1.14.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^perl@5.32.1%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu18.04-zen
            ^berkeley-db@18.1.40%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^gdbm@1.18.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^readline@8.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^ncurses@6.2%gcc@7.5.0~symlinks+termlib arch=linux-ubuntu18.04-zen
        ^xextproto@7.3.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^xtrans@1.3.5%gcc@7.5.0 arch=linux-ubuntu18.04-zen
    ^qt@4.8.7%gcc@7.5.0~dbus~debug~doc~examples~framework~gtk+gui~opengl~phonon+shared+sql+ssl+tools~webkit patches=03f9893882c63a9a9617d9ecab9caf80661c96df13f596a9b24fc4579490144a,4f40b7c6230471c19fd1aea45e264a1642c654023d2da1c8063a05ef095d6586,88ec914c9f2026599cce3258006db3fab48138cb9b37f0a7a182d6731fbe2d4d,a4f37427b34d88df6a6adc16e1e6c805087d4f3a7c658e1797450b0912fcbf5d arch=linux-ubuntu18.04-zen
        ^fontconfig@2.13.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^font-util@1.3.2%gcc@7.5.0 fonts=encodings,font-adobe-100dpi,font-adobe-75dpi,font-adobe-utopia-100dpi,font-adobe-utopia-75dpi,font-adobe-utopia-type1,font-alias,font-arabic-misc,font-bh-100dpi,font-bh-75dpi,font-bh-lucidatypewriter-100dpi,font-bh-lucidatypewriter-75dpi,font-bh-ttf,font-bh-type1,font-bitstream-100dpi,font-bitstream-75dpi,font-bitstream-speedo,font-bitstream-type1,font-cronyx-cyrillic,font-cursor-misc,font-daewoo-misc,font-dec-misc,font-ibm-type1,font-isas-misc,font-jis-misc,font-micro-misc,font-misc-cyrillic,font-misc-ethiopic,font-misc-meltho,font-misc-misc,font-mutt-misc,font-schumacher-misc,font-screen-cyrillic,font-sun-misc,font-winitzki-cyrillic,font-xfree86-type1 arch=linux-ubuntu18.04-zen
                ^autoconf@2.69%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^m4@1.4.18%gcc@7.5.0+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-ubuntu18.04-zen
                        ^libsigsegv@2.12%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^automake@1.16.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^bdftopcf@1.0.5%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^fontsproto@2.1.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^libxfont@1.5.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                        ^freetype@2.10.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                            ^bzip2@1.0.8%gcc@7.5.0+shared arch=linux-ubuntu18.04-zen
                                ^diffutils@3.7%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                                    ^libiconv@1.16%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                            ^libpng@1.6.37%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                        ^libfontenc@1.1.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^mkfontdir@1.0.7%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^mkfontscale@1.1.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^gperf@3.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libxml2@2.9.10%gcc@7.5.0~python arch=linux-ubuntu18.04-zen
                ^xz@5.2.5%gcc@7.5.0~pic arch=linux-ubuntu18.04-zen
            ^util-linux-uuid@2.36%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^glib@2.66.2%gcc@7.5.0~libmount patches=b3fd45063a19c871048aa1f28692293ab8971a871bdcbe65f06f17fdd79db9e2 tracing=none arch=linux-ubuntu18.04-zen
            ^gettext@0.21%gcc@7.5.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-zen
                ^tar@1.32%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libffi@3.3%gcc@7.5.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu18.04-zen
            ^meson@0.56.0%gcc@7.5.0 patches=7a7ff3bffb5c1996faf178da6765d578394624edb96fe9297bf21f2e3cd9ebda arch=linux-ubuntu18.04-zen
                ^cmake@3.19.2%gcc@7.5.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu18.04-zen
                    ^openssl@1.0.2u%gcc@7.5.0+systemcerts arch=linux-ubuntu18.04-zen
                ^ninja@1.10.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^python@3.8.7%gcc@7.5.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-ubuntu18.04-zen
                        ^expat@2.2.10%gcc@7.5.0+libbsd arch=linux-ubuntu18.04-zen
                        ^sqlite@3.34.0%gcc@7.5.0+column_metadata+fts~functions~rtree arch=linux-ubuntu18.04-zen
                ^py-setuptools@50.3.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^pcre@8.44%gcc@7.5.0~jit+multibyte+utf arch=linux-ubuntu18.04-zen
        ^icu4c@67.1%gcc@7.5.0 cxxstd=11 arch=linux-ubuntu18.04-zen
        ^libjpeg-turbo@2.0.6%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^nasm@2.15.05%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libmng@2.0.3%gcc@7.5.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu18.04-zen
            ^lcms@2.9%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^libtiff@4.1.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libsm@1.2.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^libice@1.0.9%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libxext@1.3.3%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libxkbcommon@0.8.2%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^bison@3.7.4%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^help2man@1.47.16%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^xkbdata@1.0.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                ^xkbcomp@1.3.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
                    ^libxkbfile@1.0.9%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^libxrender@0.9.10%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^renderproto@0.11.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^xcb-util-image@0.4.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
            ^xcb-util@0.4.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^xcb-util-keysyms@0.4.0%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^xcb-util-renderutil@0.3.9%gcc@7.5.0 arch=linux-ubuntu18.04-zen
        ^xcb-util-wm@0.4.1%gcc@7.5.0 arch=linux-ubuntu18.04-zen

and 
$ spack install silo+silex~mpi
==> silo: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/cessenat/spack-stage/spack-stage-silo-4.10.2-xadiil5d3se2jsk74dy5mtim7yq3ijwx/spack-src/configure' '--prefix=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/silo-4.10.2-xadiil5d3se2jsk74dy5mtim7yq3ijwx' '--with-hdf5=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/hdf5-1.10.7-kalzmvwon3p5x6j4477ybah6becmousp/include,/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/hdf5-1.10.7-kalzmvwon3p5x6j4477ybah6becmousp/lib' '--with-zlib=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/zlib-1.2.11-egwcgg2m23u6kiqdrywuxn5tansov3pp/include,/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/zlib-1.2.11-egwcgg2m23u6kiqdrywuxn5tansov3pp/lib' '--enable-install-lite-headers' '--enable-fortran' '--enable-silex' '--enable-shared' '--with-Qt-dir=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7' '--with-Qt-lib=QtGui -lQtCore' '--x-includes=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/libx11-1.7.0-wdieafv2ms7kxfkbi43ky6xhspdzg4y4/include' '--x-libraries=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/libx11-1.7.0-wdieafv2ms7kxfkbi43ky6xhspdzg4y4/lib'

1 error found in build log:
     196        QT_LIBS=-L/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/lib -lQtGui -lQ
            tCore  -lSM -lICE  -L/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/libx11-1.7.0-wdieafv2ms7kxfkbi43ky6xhspdzg4y4/lib 
            -lX11 -lXext -lXmu -lXt -lXi
     197        QT_UIC=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/bin/uic
     198        QT_MOC=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/bin/moc
     199        QT_LRELEASE=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/bin/lrelease
     200        QT_LUPDATE=/home2/cessenat/Softs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/qt-4.8.7-ksc4y6mx4ulvmlmpji3vy7hcrkhufkk7/bin/lupdate
     201    checking correct functioning of Qt installation... failure
  >> 202    configure: error: Failed to find matching components of a complete
     203                      Qt installation. Try using more options,
     204                      see ./configure --help.

See build log for details:
  /tmp/cessenat/spack-stage/spack-stage-silo-4.10.2-xadiil5d3se2jsk74dy5mtim7yq3ijwx/spack-build-out.txt

Reply:
> `silo depends on qt which by default uses the +gui variant`
Yep, that's right, but spack also currently "allows" building against `qt~gui` even though that will fail. I was just checking that your system was correctly using the `+gui`, and suggesting that it would be a good "safety check" to add the `+gui` variant to Silo's `qt` requirement while you're at it.

Reply:
Ok, I added +gui to the qt dependency in Silo.